Mathlib/CategoryTheory/FiberedCategory/HomLift.lean

```diff
@@ -0,0 +1,244 @@
+/-
+Copyright (c) 2024 Calle Sönne. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Lezeau, Calle Sönne
+-/
+
+import Mathlib.CategoryTheory.Functor.Category
+import Mathlib.CategoryTheory.CommSq
+
+/-!
+
+# HomLift
+
+Given a functor `p : 𝒳 ⥤ 𝒮`, this file provides API for expressing the fact that `p(φ) = f`
+for given morphisms `φ` and `f`.
+
+## Main definition
+
+Given morphism `φ : a ⟶ b` in `𝒳` and `f : R ⟶ S` in `𝒮`, `p.IsHomLift f φ` is defined as a
+structure containing the data that `p(a) = R`, `p(b) = S` and the fact that the following square
+commutes
+```
+  p(a) --p(φ)--> p(b)
+  |               |
+  |               |
+  v               v
+  R -----f------> S
+```
+where the vertical arrows are given by `eqToHom` corresponding to the equalities `p(a) = R` and
+`p(b) = S`.
+
+-/
+
+universe u₁ v₁ u₂ v₂
+
+open CategoryTheory Category
+
+variable {𝒮 : Type u₁} {𝒳 : Type u₂} [Category 𝒳] [Category 𝒮]
+
+namespace CategoryTheory
+
+/-- The proposition that an arrow a --φ--> b lifts an arrow R --f--> S in 𝒮 via p. This is
+often drawn as:
+```
+  a --φ--> b
+  -        -
+  |        |
+  v        v
+  R --f--> S
+``` -/
+class Functor.IsHomLift (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) : Prop where
+  domain_eq : p.obj a = R := by aesop_cat
+  codomain_eq : p.obj b = S := by aesop_cat
+  fac : f = eqToHom domain_eq.symm ≫ p.map φ ≫ eqToHom codomain_eq := by aesop_cat
+
+namespace IsHomLift
+
+/-- For any arrow `φ : a ⟶ b` in `𝒳`, `φ` lifts the arrow `p.map φ` in the base `𝒮`-/
+instance self (p : 𝒳 ⥤ 𝒮) {a b : 𝒳} (φ : a ⟶ b) : p.IsHomLift (p.map φ) φ where
+instance (p : 𝒳 ⥤ 𝒮) (a : 𝒳) : p.IsHomLift (𝟙 (p.obj a)) (𝟙 a) where
+
+@[simp]
+protected lemma id {p : 𝒳 ⥤ 𝒮} {R : 𝒮} {a : 𝒳} (ha : p.obj a = R) : p.IsHomLift (𝟙 R) (𝟙 a) where
+
+lemma domain_eq (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b)
+    [p.IsHomLift f φ] : p.obj a = R :=
+  Functor.IsHomLift.domain_eq f φ
```

Comment:
As the first five lemmas have the same parameters, you could define them in a section:

```suggestion

section

variable (p : 𝒳 ⥤ 𝒮) {R S : 𝒮} {a b : 𝒳} (f : R ⟶ S) (φ : a ⟶ b) [p.IsHomLift f φ]

lemma domain_eq  : p.obj a = R :=
  Functor.IsHomLift.domain_eq f φ
```

Reply:
Also, `p : 𝒳 ⥤ 𝒮` could be defined as a variable more globally in this file.